_posts/2015-07-15-testautomation-allergy.md

```diff
@@ -0,0 +1,60 @@
+
+Test automation should be used to automate *checks*. Assert(A==A). Then add flesh to the bones. And then you have an automated check. The moment you start using test automation to do anything more complex, you will be pouring money into something that might never pay it back. Use test automation as a tool to make development easier (and at some point even help testing). You don't want to check all the things every time, but you want to test how things actually work. You don't want the stuff to break that (we think) worked previously because you don't want regression. Even the best test automation will not reveal everything, as we all know.
+
+Of course there are different granularities in test automation. From unit-testing individual classes and methods to full blown user simulations. And they all have some importance to some task in developing software. Not always but often enough, people talk about test automation as a replacement for all other testing. They try to automate testing and then happily watch the machine do all the work. Sadly, test automation cannot perform the following check:
+
+"Check that nothing happens that shouldn’t happen and everything else happens that should happen for all variations of this scenario and all possible states of systems and all possible visual/audio/tangible things user might observe and all possible states of the backend and all possible states of the system as a whole, and anything happening in the surrounding systems that should not matter but might matter."
+
+A human cannot test that on one go, but I bet my year’s salary that humans fare better at the task than a machine. Of course contexts vary and there might be occasions where a test automation can test absolutely everything. I (nor anyone whose opinion I value) have ever seen that.
+
+
+The framework of integrity and constant improvement
+----------------------------------------------------
+
+Why the f*** do we want to repeat a test? Why the f*** do we want to repeat a test? Why the f*** do we want to repeat a test? Why the f*** do we want to repeat a test? 
+
+See how silly it is? If you think of it, there are many variables we don't know. We have randomness in our servers, networks, everywhere! Even bloody electrical current isn't stable. So why on earth do we want to try and repeat the situation over and over, if the world we live in is not repeating itself? You might want to answer "to verify that regression doesn't exist". Isn't the point and purpose of regression testing to see if there is something broken? Shouldn't we be searching for that thing that MIGHT be broken?
```

Comment:
"If you think of it, there are many variables we don't know." => "If you think about it, there are many variables we don't know."

"Even bloody electrical current isn't stable" => "Even the bloody electrical current isn't stable"

"Shouldn't we be searching for that thing that MIGHT be broken?" => "Shouldn't we be searching for the thing that MIGHT be broken?"